File: longLine/try1.py
#!/usr/bin/env python3
# encoding: utf-8

def parser():
    while 1:
        data = list(input().split(' '))
        for number in data:
            if len(number) > 0:
                yield(number)   

input_parser = parser()

def get_word():
    global input_parser
    return next(input_parser)

def get_number():
    data = get_word()
    try:
        return int(data)
    except ValueError:
        return float(data)

n = get_number()
h = get_number()
a = get_number()
c = get_number()
q = get_number()
sum = 0
stack = []

stack.append(h)

for i in range(n-1):
	h = (a*h+c)%q
	sum += len(stack)
	while(len(stack) != 0):
		if(stack[-1] > h):
			break
		else:
			stack.pop()
	stack.append(h)

print(sum)

File: longLine/try1.cpp
#include <iostream>
#include <stack>
using namespace std;

int next(int a, int h, int c, int q){
	return (a*h+c)%q;
}

int main(){
	int n, h, q, a, c;
	int sum = 0; 
	stack<int> sol; 

	cin >> n >> h >> a >> c >> q;
	sol.push(h);
	for(int i = 1; i < n; ++i){
		h = next(a, h, c, q)
		sum += sol.size();
		while(!sol.empty()){
			// cout << sol.top() << ' ';
			if(sol.top() > h){
				break;
			}
			else{
				sol.pop();
			}
		}
		// cout << endl;
		sol.push(h);
	}

	cout << sum << endl;

}